internal/backend/temporalclient/largepayloads_codec.go

```diff
@@ -0,0 +1,178 @@
+	"go.autokitteh.dev/autokitteh/sdk/sdktypes"
+)
+
+const (
+	largeObjectKeyMetadataKey  = "ak_lo_key"
+	largeObjectSizeMetadataKey = "ak_lo_size"
+)
+
+type LargePayloadConfig struct {
+	StoreType string `koanf:"store_type"` // "none" to disable.
+
+	// ThresholdSize is the maximum size of a payload that will be stored in Temporal.
+	// If a payload is larger than this size, it will be stored in a separate store.
+	ThresholdSize int `koanf:"threshold_size"`
+
+	FileStoreRootPath string `koanf:"file_store_root_path"` // if store_type == "file".
+	S3StoreBucketName string `koanf:"s3_store_bucket_name"` // if store_type == "s3".
+}
+
+type largePayloadCodec struct {
+	l     *zap.Logger
+	cfg   LargePayloadConfig
+	store gokv.Store
+}
+
+func newLargePayloadCodec(l *zap.Logger, cfg LargePayloadConfig) (converter.PayloadCodec, error) {
+	var store gokv.Store
+
+	switch cfg.StoreType {
+	case "":
+		return nil, errors.New("store type not specified")
+	case "none":
+		l.Warn("large payloads store is disabled")
+		return nil, nil
+	case "inmem":
+		store = gomap.NewStore(gomap.Options{Codec: encoding.Gob})
+		l.Warn("using in-memory store for large payloads")
+	case "file":
+		var (
+			err error
+			ext = "ext"
+		)
+
+		dir := cfg.FileStoreRootPath
+		if dir == "" {
+			dir = filepath.Join(xdg.DataHomeDir(), "temporal-large-payloads")
+		}
+
+		if store, err = file.NewStore(file.Options{
+			Codec:             encoding.Gob,
+			FilenameExtension: &ext,
+			Directory:         dir,
+		}); err != nil {
+			return nil, err
+		}
+
+		l.Info("using file store for large payloads", zap.String("root", dir))
+	case "s3":
+		var err error
+		if store, err = s3.NewClient(s3.Options{
+			BucketName: cfg.S3StoreBucketName,
+			Codec:      encoding.Gob,
+		}); err != nil {
+			return nil, err
+		}
+
+		l.Info("using S3 store for large payloads", zap.String("bucket", cfg.S3StoreBucketName))
+	default:
+		return nil, errors.New("unknown store type")
+	}
+
+	return &largePayloadCodec{cfg: cfg, l: l, store: store}, nil
+}
+
+func (c *largePayloadCodec) Encode(payloads []*commonpb.Payload) ([]*commonpb.Payload, error) {
+	out := make([]*commonpb.Payload, 0, len(payloads))
+
+	for _, in := range payloads {
+		data := in.GetData()
+		size := len(data)
+
+		l := c.l.With(zap.Int("size", size))
+
+		if size < c.cfg.ThresholdSize {
+			l.Debug("payload is below threshold size, storing in Temporal")
+			out = append(out, in)
+			continue
+		}
+
+		key := sdktypes.NewUUID().String()
+		l = l.With(zap.String("key", key))
+
+		l.Debug("payload is above threshold size, storing in external store")
+
+		if err := c.store.Set(key, in.GetData()); err != nil {
+			l.Error("failed to store payload in external store", zap.Error(err))
+			return nil, err
+		}
+
+		md := maps.Clone(in.Metadata)
+		if md == nil {
+			md = make(map[string][]byte, 2)
+		}
+		md[largeObjectKeyMetadataKey] = []byte(key)
+		md[largeObjectSizeMetadataKey] = []byte(strconv.Itoa(size))
```

Comment:
Maybe also add a checksum? OK to leave as a distant TODO (but with a ticket ID)

Reply:
added sha.